docs/development/etcdoperatortask-design.md

```diff
@@ -0,0 +1,501 @@
+            Description: fmt.Sprintf("Failed to create HTTP request for %s", url),
+            Error:       err,
+        }
+    }
+    o.logger.Info("Triggering snapshot", "url", url)
+
+    resp, err := o.httpClient.Do(req)
+    if err != nil {
+        return &operatortask.TaskResult{Description: "Snapshot HTTP request failed", Error: err}
+    }
+    defer resp.Body.Close()
+    if resp.StatusCode != 200 {
+        return &operatortask.TaskResult{Description: fmt.Sprintf("Snapshot failed, status: %s", resp.Status), Completed: false}
+    }
+    return &operatortask.TaskResult{Description: "Triggered snapshot on pod", Completed: true}
+}
+
+func (o *OnDemandSnapshot) Admit(ctx context.Context) *operatortask.TaskResult {
+    // Implement the admission logic (e.g., check if Etcd is ready)
+    etcd := &v1alpha1.Etcd{}
+    if err := o.client.Get(ctx, o.etcdReference, etcd); err != nil {
+        return &operatortask.TaskResult{Description: "Failed to get Etcd resource", Error: err}
+    }
+    if etcd.Status.ReadyReplicas < 1 {
+        return &operatortask.TaskResult{
+            Description: "No ready replicas for Etcd",
+            Error: druiderr.WrapError(nil,
+                ERR_PRECONDITION_CHECK_ON_DEMAND_SNAPSHOT,
+                operationAdmit,
+                fmt.Sprintf("No ready replicas for Etcd: %v", o.etcdReference)),
+            Completed: true,
+        }
+    }
+    return &operatortask.TaskResult{Description: "Admission successful", Completed: true}
+}
+
+func (o *OnDemandSnapshot) Cleanup(ctx context.Context) *operatortask.TaskResult {
+    // No-op for snapshot tasks
+    return &operatortask.TaskResult{Description: "Cleanup done", Completed: true}
+}
+
+
+
+```
+</details>
+
+
+#### Example YAML for an On-Demand Snapshot Task
+
+```yaml
+apiVersion: druid.gardener.cloud/v1alpha1
+kind: EtcdOperatorTask
+metadata:
+  name: on-demand-snapshot-task
+  namespace: default
+spec:
+  type: OnDemandSnapshot
+  etcdRef:
+    name: etcd-test
+    namespace: default
+  config:
+    snapshotType: "full"
+    timeoutSeconds: 60
+  ttlSecondsAfterFinished: 600
+```
+
+
+### Registering a New OperatorTask
+
+1. Implement the `TaskHandler` and its constructor.
+2. Add the new handler to `createTaskHandlerInstance`:
+   ```go
+   func (r *Reconciler) createTaskHandlerInstance(task *v1alpha1.EtcdOperatorTask) (tasks.TaskHandler) {
+       switch task.Spec.Type {
+       case v1alpha1.<NewTaskType>:
+           return <NewTaskType>.New(r.client, r.logger, task)
+       // Add more cases for other task types
+       default:
+       }
+   }
+   ```
+
+**Benefits:**
+- **Extensible:** New OperatorTasks can be added without modifying the core controller logic.
+- **Decoupled:** OperatorTask implementations are independent from the reconciler logic.
+
+
+
+---
+
+## Next Steps
+
+- Design and implement HTTP endpoints in the etcd-backup-restore server to expose the etcd maintenance API, supporting both sync and async operations.
```

Comment:
When you say async, you're talking about doing it via an API resource that etcdbr reads and then performs the snapshot upload and record the status result to?

Reply:
Current snapshot API is not fully RESTful and does not support async operations. As part of this design, we plan to refactor the snapshot API to follow REST principles and support both synchronous and asynchronous workflows. The goal is to provide standard HTTP(S) endpoints to trigger operations and poll their status, making it easier to integrate and automate etcd maintenance tasks. The exact API shape is yet to be designed.